Allow managing persistence configurations and enable filters

J-N-K opened the pull request:
Depends on #2994 
Fixes #3590 

The aim of this PR is to allow managing persistence service configurations via the REST API (UI).

To achieve this
- A `PersistenceServiceConfigurationRegistry` has been introduced. It needs to add an own interface for the change listener because one class can't inherit two `RegistryChangeListener<T>` with different types and the persistence manager needs to listen to both, the `ItemRegistry` and the `PersistenceServiceConfigurationRegistry`.
- The `PersistenceManager` interface has been removed.
- The `PersistenceManagerImpl` has been renamed to `PersistenceManager` and makes use of the new registry.
- The `PersistenceModelManager` has been refactored to a `PersistenceServiceConfigurationProvider`.
- A `ManagedPersistenceServiceConfigurationProvider` has been introduced.
- The core REST bundle has been extended to allow adding configurations. Unmanaged configurations can also be retrieved. 

Signed-off-by: Jan N. Klug <github@klug.nrw>

J-N-K commented:
@ghys Can you have a look at the REST part and let me know if that is a good design for the UI?

J-N-K commented:
I'll add more extensive test coverage to ensure that my refactoring are all correct.

splatch commented:
Since you touch fairly important base - have a look on `PersistenceFilter` which is blind interface without use. While it can be referred in persistence service config it has no use. I made it an `Predicate<Item>` which allows to use it as an exclusion filter for certain items. Obviously there is still a problem of mapping the filters, but that's other thing. Other elephant then would be the xtext part which would need to be provided by someone who knows how to work with it (or omitted as deprecated).

J-N-K commented:
@splatch I enabled the `PersistenceFilter`.

splatch commented:
> @splatch I enabled the `PersistenceFilter`.

Yes, I saw it and see you use it to throttle or limit writes to persistence stuff. It makes a lot of sense cause this can be used when multiple stores are used. My approach so far was focusing on use of profiles which worked for single persistence but not for all.

I think there is still an area for improvement in excluding certain items, cause `mapFilter` is currently fairly limited. Its really about use cases reported several times in forums (where I advertised own externalized XML config for that) to serve cases where it is easier to mark item with an tag than create a completely new group without just few members.
Knowing the limit of xtext I think it could be done with a basic `instanceof` call to see if ie. incoming filter instance is already compatible with contract and does not require further mapping.

J-N-K commented:
The two filters I added were already defined in the model, that's why I implemented them. If someone knows how to properly extend that, more filters can be added. Maybe something like `!itemName` could be used to exclude items. But it's difficult if we need wildcards. This should ne done in a separate PR. I only added the filters here because I needed to figure out how to model them for the REST API.

splatch commented:
@J-N-K I'm perfectly fine with your work, my stuff is sitting on OH 3.0.x fork, so I still have plenty of time to fight with 3.3 updates. ;-)

J-N-K commented:
@florian-h05 FYI. This is also lacking UI support and I don't have a good idea how this should look like.

florian-h05 commented:
Thanks for the hint, I have also seen https://github.com/openhab/openhab-webui/issues/1463.
I am currently super busy wit my final exams (Abitur), but I put this on my UI list 👍

clinique commented:
> Abitur

May the force be with you !

J-N-K commented:
> I am currently super busy wit my final exams (Abitur), but I put this on my UI list 👍

Best of luck!

kaikreuzer reviewed (changes requested):
Many thanks! I have just a few small comments below.

J-N-K commented on bundles/org.openhab.core.model.persistence/src/org/openhab/core/model/persistence/internal/PersistenceModelManager.java:
```diff
@@ -47,75 +59,55 @@
     protected void deactivate() {
         modelRepository.removeModelRepositoryChangeListener(this);
-        for (String modelName : modelRepository.getAllModelNamesOfType("persist")) {
-            removeModel(modelName);
```
Technically it is not needed as long as the registry is the only listener, because on de-activation the provider is removed from the registry and the implementation in `AbstractRegistry` then removes all elements that were provided by the removed provider.

But if you prefer to remove it anyway, I can re-add it.

J-N-K replied:
After thinking about it, I believe that it is cleaner if we remove it ourselves.

J-N-K commented on bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/FilterCriteria.java:
```diff
@@ -65,13 +68,13 @@ public enum Ordering {
 
-    /** filter result to only contain entries that are equal to or after the given datetime */
-    private ZonedDateTime beginDate;
+    /** filter result to only contain entries that are newer than the given date */
```
It was changed from "newer than" to "equal to or after" in #3164 and I missed that when rebasing.

J-N-K commented on bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/filter/PersistenceFilter.java:
```diff
@@ -0,0 +1,85 @@
+import org.openhab.core.items.Item;
+
+/**
+ *
```
No, it was missing. I have added it now.

kaikreuzer reviewed (approved):
Thanks again!

lolodomo commented:
> Other elephant then would be the xtext part which would need to be provided by someone who knows how to work with it (or omitted as deprecated).

@J-N-K : What exactly is missing in current XText for persistence? 
I can try to add/fix it if I understand what is the problem.

J-N-K commented:
Filters are already enabled (I think we need documentation for that). Currently there are two types of filters: value-based and time-based. If we would add more, then there would also be the need for XText support, but I think currently we are fine.

lolodomo commented:
Fine. 
I see them in XText syntax.
